cmds/core/wget/wget_test.go

```diff
@@ -40,131 +43,147 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
-var tests = []struct {
-	name    string
-	flags   []string // in, %[1]d is the server's port, %[2] is an unopen port
-	url     string   // in
-	content string   // out
-	retCode int      // out
-}{
-	{
-		name:    "basic",
-		flags:   []string{},
-		url:     "http://localhost:%[1]d/200",
-		content: content,
-		retCode: 0,
-	},
-	{
-		name:    "ipv4",
-		flags:   []string{},
-		url:     "http://127.0.0.1:%[1]d/200",
-		content: content,
-		retCode: 0,
-	},
-	// TODO: CircleCI does not support ipv6
-	// {
-	// 	name:    "ipv6",
-	// 	flags:   []string{},
-	// 	url:     "http://[::1]:%[1]d/200",
-	// 	content: content,
-	// 	retCode: 0,
-	// },
-	{
-		name:    "redirect",
-		flags:   []string{},
-		url:     "http://localhost:%[1]d/302",
-		content: "",
-		retCode: 0,
-	},
-	{
-		name:    "4xx error",
-		flags:   []string{},
-		url:     "http://localhost:%[1]d/404",
-		content: "",
-		retCode: 1,
-	},
-	{
-		name:    "5xx error",
-		flags:   []string{},
-		url:     "http://localhost:%[1]d/500",
-		content: "",
-		retCode: 1,
-	},
-	{
-		name:    "no server",
-		flags:   []string{},
-		url:     "http://localhost:%[2]d/200",
-		content: "",
-		retCode: 1,
-	},
-}
-
-func getListener(t *testing.T) (net.Listener, int) {
+// unusedPort return a port number and setup a listener that closes connection
+// for any request
+func unusedPort(t *testing.T) (int, func() error) {
 	t.Helper()
-	l, err := net.Listen("tcp", ":0")
+	a, err := net.ResolveTCPAddr("tcp", ":0")
 	if err != nil {
-		t.Fatalf("error setting up TCP listener: %v", err)
+		t.Fatal(err)
+	}
+	l, err := net.ListenTCP("tcp", a)
+	if err != nil {
+		t.Fatal(err)
 	}
-	return l, l.Addr().(*net.TCPAddr).Port
-}
 
-// TestWget implements a table-driven test.
-func TestWget(t *testing.T) {
-	// Start a webserver on a free port.
-	l, port := getListener(t)
-	defer l.Close()
-	ul, unusedPort := getListener(t)
-	defer ul.Close()
 	go func() {
 		for {
-			conn, err := ul.Accept()
+			conn, err := l.Accept()
 			if err != nil {
-				// End of test.
 				return
 			}
 			conn.Close()
 		}
 	}()
 
-	h := handler{}
-	go func() {
-		log.Print(http.Serve(l, h))
-	}()
+	return l.Addr().(*net.TCPAddr).Port, l.Close
+}
+
+// TestWget implements a table-driven test.
+func TestWget(t *testing.T) {
+	srv := httptest.NewServer(handler{})
+	defer srv.Close()
+	tmpDir := t.TempDir()
+	freePort, closer := unusedPort(t)
+	defer closer()
+
+	var tests = []struct {
+		name        string
+		url         string // in
+		wantContent string // out
+		outputPath  string
+		err         error
+	}{
+		{
+			name:        "ipv4",
+			url:         fmt.Sprintf("%s/200", srv.URL),
+			wantContent: content,
+			outputPath:  filepath.Join(tmpDir, "basic"),
```

Comment:
instead of doing `filepath.Join(tmpDir, ...)` in each test case, maybe you can move the `tmpDir := t.TempDir()` and the `filepath.Join` into the test loop code?

Reply:
updated